xdelta3-dir-patcher

```diff
@@ -21,11 +56,730 @@ class AttributeDict(dict):
+                #      between archive<->archive to copy these values
+
+                # print('WARNING! Could not change uid/gid of', target_path)
+                pass
+
+    def create(self, base_dir):
+        if path.isdir(self.path):
+            raise Exception('Error! Archive already present!')
+
+        copytree(base_dir, self.path,
+                 symlinks = True,
+                 ignore_dangling_symlinks = False)
+
+class XDelta3TarImpl(XDelta3AbstractArchiveImpl):
+    TAR_FORMAT = 'gz'
+
+    def __init__(self, archive_path, for_writing = False):
+        super().__init__()
+
+        self._items = None
+        flags = 'r:*'
+
+        if for_writing:
+            if path.isfile(archive_path):
+                raise Exception('Error! Archive already present!')
+
+            flags = 'w:%s' % self.TAR_FORMAT
+
+        self.archive_object = tarfile.open(archive_path, flags)
+        self.archive_name = path.basename(archive_path)
+
+        # Pre-fetch member data to ensure only one thread tries to create
+        # the initial list and to allow further listings to not need locks
+        if not for_writing:
+            self.list_items()
+
+    def _close_archive(self):
+        self.archive_object.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._close_archive()
+
+    def close(self):
+        self._close_archive()
+
+    @staticmethod
+    def can_open(archive):
+        return path.isfile(archive) and tarfile.is_tarfile(archive)
+
+    def _add_listing_object(self, dir_listing, method, member):
+        setter_func = getattr(dir_listing, method)
+
+        file_obj = setter_func(path.basename(member.name),
+                               member,
+                               member.mode,
+                               member.uname,
+                               member.uid,
+                               member.gname,
+                               member.gid,
+                               member.issym())
+
+        if file_obj and file_obj.is_link:
+            file_obj.link_target = member.linkname
+
+        return file_obj
+
+    def _create_dir_structure_to(self, items, target_path):
+        target_path_segments = target_path.split(path.sep)
+        for index, segment in enumerate(target_path_segments):
+            segment_path = path.sep.join(target_path_segments[0:index + 1])
+
+            # Skip paths that are in hierarchy
+            if segment_path in items:
+                continue
+
+            parent_obj = items[None]
+            if index != 0:
+                parent_obj = items[path.dirname(segment_path)]
+
+            subdir_obj = DirListing(segment)
+            parent_obj.add_subdir(subdir_obj)
+
+            items[segment_path] = subdir_obj
+
+    def _safe_makedirs(self, target_dir):
+        super()._acquire_lock()
+        makedirs(target_dir, exist_ok = True)
+        super()._release_lock()
```

Comment:
@sgnn7 I think this got lost because the commit is outdated, but why is locking needed around the makedirs? I would expect that exist_ok would make that unnecessary.